extensions/amp-sidebar/1.0/sidebar-animations-hook.js

```diff
@@ -0,0 +1,192 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Side, useValueRef} from './sidebar-config';
+import {assertDoesNotContainDisplay, setStyles} from '../../../src/style';
+import {useLayoutEffect, useRef} from '../../../src/preact';
+
+const ANIMATION_DURATION = 350;
+const ANIMATION_EASE_IN = 'cubic-bezier(0,0,.21,1)';
+
+const ANIMATION_KEYFRAMES_FADE_IN = [{'opacity': '0'}, {'opacity': '1'}];
+const ANIMATION_KEYFRAMES_SLIDE_IN_LEFT = [
+  {'transform': 'translateX(-100%)'},
+  {'transform': 'translateX(0)'},
+];
+const ANIMATION_KEYFRAMES_SLIDE_IN_RIGHT = [
+  {'transform': 'translateX(100%)'},
+  {'transform': 'translateX(0)'},
+];
+
+const ANIMATION_STYLES_SIDEBAR_LEFT_INIT = {'transform': 'translateX(-100%)'};
+const ANIMATION_STYLES_SIDEBAR_RIGHT_INIT = {'transform': 'translateX(100%)'};
+const ANIMATION_STYLES_BACKDROP_INIT = {'opacity': '0'};
+const ANIMATION_STYLES_SIDEBAR_FINAL = {'transform': ''};
+const ANIMATION_STYLES_BACKDROP_FINAL = {'opacity': ''};
+
+/** @private @enum {number} */
+const Direction = {
+  OPENING: 0,
+  CLOSING: 1,
+  NOT_ANIMATING: 2,
+};
+
+/**
+ * @param {!Element} element
+ * @param {!Object<string, *>} styles
+ */
+function safelySetStyles(element, styles) {
+  setStyles(element, assertDoesNotContainDisplay(styles));
+}
+
+/**
+ * @param {boolean} opened
+ * @param {{current: function|undefined}} onAfterClose
+ * @param {string|undefined} side
+ * @param {{current: Element|null}} sidebarRef
+ * @param {{current: Element|null}} backdropRef
+ * @param {function} setMounted
+ */
+export function useSidebarAnimation(
+  opened,
+  onAfterClose,
+  side,
+  sidebarRef,
+  backdropRef,
+  setMounted
+) {
+  const onAfterCloseRef = useValueRef(onAfterClose);
+  const sidebarAnimationRef = useRef(null);
+  const backdropAnimationRef = useRef(null);
+  const animationDirectionRef = useRef(Direction.NOT_ANIMATING);
+  useLayoutEffect(() => {
+    const sidebarElement = sidebarRef.current;
+    const backdropElement = backdropRef.current;
+    if (!sidebarElement || !backdropElement || !side) {
+      return;
+    }
+
+    const postVisibleAnim = () => {
+      safelySetStyles(sidebarElement, ANIMATION_STYLES_SIDEBAR_FINAL);
+      safelySetStyles(backdropElement, ANIMATION_STYLES_BACKDROP_FINAL);
+      sidebarAnimationRef.current = null;
+      backdropAnimationRef.current = null;
+      animationDirectionRef.current = Direction.NOT_ANIMATING;
+    };
+    const postInvisibleAnim = () => {
+      if (onAfterCloseRef.current) {
+        onAfterCloseRef.current();
+      }
+      sidebarAnimationRef.current = null;
+      backdropAnimationRef.current = null;
+      animationDirectionRef.current = Direction.NOT_ANIMATING;
+      setMounted(false);
+    };
+
+    // currently animating
+    if (animationDirectionRef.current !== Direction.NOT_ANIMATING) {
+      if (
+        (opened && animationDirectionRef.current === Direction.CLOSING) ||
```

Comment:
Is it possible to fall into the `else` block here? In other words, can the sidebar ever be `open && opening`, or `!open && closing`? 

If so, I think we should still keep the `cancel()` (which I see were removed from cleanup callback) for both animations in the `else`. If not, can we simplify the `OPENING` and `CLOSING` states to a single `ANIMATING`, `IN_PROGRESS`, or similar?

Reply:
It is possible to fall into the `else` block  here.  So when user does `api.open()` corresponds with `opened` and `api.close()` corresponds to `!opened`.  (api is the imperative API for the component).

So it's possible for the user to attempt to open the sidebar while it's already opening (or try to close when it's already closing).  In my opinion we should not cancel here, since the user experience is not that smooth if the sidebar automatically snaps shut or snaps open.

Additionally, we will need to track the animation direction so am unable to simplify to `ANIMATING`, `NOT_ANIMATING` (at least not in the current implementation).

I wish we could though and open to other suggestions :)

Reply:
Update -> 

We can only enter this `effect` when the user changes animation direction.  So we do not need to track animation direction.  So now simplifying the logic to check whether or not the animation is currently happening or not in the latest revision.